### PR TITLE
Add gpuCI GPU testing, automatic GPU detection, and support for more compilers

### PR DESCRIPTION
**Pull request by brycelelbach**
Add gpuCI GPU testing, automatic GPU detection, and support for more compilers:

* CMake: Add support for detecting the compute archs of the GPUs in your system
  at configure time.
* gpuCI: Add a GPU node configuration that builds and tests as little as possible.
* gpuCI: Cleanup logic for different build and test configurations.
* gpuCI: Fix an unfortunate typo in `determine_build_parallelism.bash` which
  led to the parallelism level not being set.
* gpuCI: Add support for NVC++.
* gpuCI: Update to CUDA 11.1 and Ubuntu 20.04.
* gpuCI: Add NVC++ and ICC configurations to the CPU axis file.
* gpuCI: Add a GPU axis file.
* gpuCI: Increase the desired memory per build thread to 4GB.
* gpuCI: Add a -j switch which controls build parallelism to `ci/local/build.bash`.

### CODE REVIEW

**Comment by brycelelbach** [comment]
ok to test

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

File: ci/common/build.bash
Comment (allisonvacanti) [inline ci/common/build.bash:91]: Per the comments on NVIDIA/cub#244, this may need to be changed to a couple of booleans.

File: cmake/detect_compute_archs.cu
Comment (allisonvacanti) [inline cmake/detect_compute_archs.cu:34]: Doesn't look like this case is handled.

File: cmake/ThrustCudaConfig.cmake
Comment (allisonvacanti) [inline cmake/ThrustCudaConfig.cmake:79]: This naming convention tho

**Review by allisonvacanti — APPROVED** [review]
LGTM, a couple potential issues to be aware of pointed out inline. We should handle the case where no arches are detected before merging, but otherwise 👍

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
Marking as "request changes" to block merging, since the "no GPUs detected" case is probably very important on the CI nodes.

File: cmake/ThrustCudaConfig.cmake
Comment (brycelelbach) [reply cmake/ThrustCudaConfig.cmake:79]: Want me to change it?

Comment (allisonvacanti) [reply cmake/ThrustCudaConfig.cmake:79]: Nah, I'm just amused by it 🙃

File: ci/common/build.bash
Comment (brycelelbach) [reply ci/common/build.bash:102]: Done.

File: cmake/detect_compute_archs.cu
Comment (brycelelbach) [reply cmake/detect_compute_archs.cu:34]: Done.

**Comment by brycelelbach** [comment]
Okay, the "ALL" case is handled now.

**Comment by brycelelbach** [comment]
ok to test

File: ci/common/build.bash
Comment (allisonvacanti) [reply ci/common/build.bash:102]: Still needs to be updated, `-DCUB_TEST_VARIANT_MINIMAL=ON`

File: cmake/detect_compute_archs.cu
Comment (allisonvacanti) [reply cmake/detect_compute_archs.cu:34]: The case where no arches are detected and the script prints "ALL" is still not handled.

**Review by allisonvacanti — COMMENTED** [review]
Is github up to date with your local branch? Some changes seem to be missing.

**Comment by brycelelbach** [comment]
I've pushed the latest changes.

**Review by allisonvacanti — APPROVED** [review]
Looks great, can't wait to try it out! 🥳

**Comment by brycelelbach** [comment]
I'm gonna wait for @raydouglass to explicitly sign-off because this will actually turn on more configs.

File: ci/axis/cpu.yml
Comment (raydouglass) [inline ci/axis/cpu.yml:7]: Can these be `cuda` and `nvhpc`? I don't have a way in Jenkins to split this out before the docker image is pulled.

The `cccl` side can be left as is. It's just the thrust side and how Jenkins will pull the docker image that's an issue.

**Review by raydouglass — CHANGES_REQUESTED** [review]

File: ci/axis/cpu.yml
Comment (brycelelbach) [reply ci/axis/cpu.yml:7]: Ah yah I can do that, no problem. I'll change them.

Comment (brycelelbach) [reply ci/axis/cpu.yml:7]: Does it help it make the change in `cccl` as well?

Comment (raydouglass) [reply ci/axis/cpu.yml:7]: The consistency is probably worth it

Comment (brycelelbach) [reply ci/axis/cpu.yml]: Fixed here.

Comment (brycelelbach) [reply ci/axis/cpu.yml]: Done in `cccl` as of https://github.com/NVIDIA/cccl/commit/d6134ad5cdfd5aa37c5209906c5acf0fdebe7d3b

**Comment by raydouglass** [comment]
run tests

I've enabled the GPU build, so let's see how it goes.

File: ci/axis/gpu.yml
Comment (raydouglass) [inline ci/axis/gpu.yml:7]: Should just be `cuda`
```suggestion
  - cuda
```

File: ci/axis/gpu.yml
Comment (raydouglass) [inline ci/axis/gpu.yml:16]: This doesn't match cccl: https://github.com/NVIDIA/cccl/blob/main/ci/axis/docker.yml#L19
```suggestion
  - 20.04
```

**Review by raydouglass — CHANGES_REQUESTED** [review]

File: ci/axis/gpu.yml
Comment (brycelelbach) [reply ci/axis/gpu.yml:7]: Fixed.

File: ci/axis/gpu.yml
Comment (brycelelbach) [reply ci/axis/gpu.yml]: Indeed it does not. Fixed.

**Comment by brycelelbach** [comment]
run tests

**Comment by raydouglass** [comment]
run tests

**Comment by brycelelbach** [comment]
Looks like the docker image pull was fixed. Builds still failed due to a tiny mistake I made in one of the scripts. Let's try it again...

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
There was a bug in the images - `${CUDACXX}` wasn't getting passed through. I'll retrigger once they're rebuilt.

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
GPU tests are still failing - for some reason, the NVIDIA runtime doesn't seem to be enabled by gpuCI. That will have to wait until after the break.

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Comment by brycelelbach** [comment]
run tests

**Pull request merged**
